src/org/mozilla/javascript/NativeString.java

```diff
@@ -941,6 +946,47 @@ protected int findPrototypeId(Symbol k)
         return 0;
     }
 
+    /**
+     * <h1>String.raw (template, ...substitutions)</h1>
+     * <p>22.1.2.4 String.raw [Draft ECMA-262 / April 28, 2021]</p>
+     */
+    private static CharSequence js_raw(Context cx, Scriptable scope, Object[] args) {
+        final Object undefined = Undefined.instance;
+        /* step 1-2 */
+        Object arg0 = args.length > 0 ? args[0] : undefined;
+        Scriptable cooked = ScriptRuntime.toObject(cx, scope, arg0);
+        /* step 3 */
+        Object rawValue = cooked.get("raw", cooked);
+        if (rawValue == NOT_FOUND) rawValue = undefined;
+        Scriptable raw = ScriptRuntime.toObject(cx, scope, rawValue);
+        /* 4-5 */
+        long literalSegments = NativeArray.getLengthProperty(cx, raw);
+        if (literalSegments == 0) return "";
+        /* step 6-7 */
+        StringBuilder elements = new StringBuilder();
+        long nextIndex = 0;
+        for (;;) {
+            /* step 8 a-i */
+            Object next;
+            if (nextIndex > Integer.MAX_VALUE) {
+                next = raw.get(Long.toString(nextIndex), raw);
```

Comment:
Would you mind looking to see if there are any other tests, perhaps in the v8 project, that we could use to get better coverage? For example, I see that this case (of the index being greater than a 32-bit int) isn't being tested.

Reply:
Can I actually just change nextIndex from a long to an int, and get rid of that step all together? It will be impossible to hit that number, because `args` would have to be an `Object[]` with `Integer.MAX_VALUE+1` elements.

Reply:
Hmm, I guess raw can have a length up to `2^53-1`, and if that happens, it will stop doing substitutions from args and use empty strings for the remainder of the substitutions. Do we really want a test that has to iterate that many times? It will probably cause an out of memory error.

Reply:
I think it would get an out of memory error because either raw would have to have a massive amount of properties, or it is sparse, and the output will repeat "undefined" for each index that doesn't have an element.

Reply:
just to confirm, I ran the following in a shell
```js
Rhino 1.7.14-SNAPSHOT 2021 05 05
js> String.raw({raw:{length:5}})
undefinedundefinedundefinedundefinedundefined
js> String.raw({raw:{length:Math.pow(2,32)+1}})
java.lang.OutOfMemoryError: Java heap space
        at java.base/java.util.Arrays.copyOf(Arrays.java:3745)
        at java.base/java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:172)
        at java.base/java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:538)
        at java.base/java.lang.StringBuilder.append(StringBuilder.java:174)
        at org.mozilla.javascript.NativeString.js_raw(NativeString.java:1145)
        at org.mozilla.javascript.NativeString.execIdCall(NativeString.java:398)
        at org.mozilla.javascript.IdFunctionObject.call(IdFunctionObject.java:100)
        at org.mozilla.javascript.optimizer.OptRuntime.call1(OptRuntime.java:45)
        at org.mozilla.javascript.gen._stdin__2._c_script_0(Unknown Source)
        at org.mozilla.javascript.gen._stdin__2.call(Unknown Source)
        at org.mozilla.javascript.ContextFactory.doTopCall(ContextFactory.java:415)
        at org.mozilla.javascript.ScriptRuntime.doTopCall(ScriptRuntime.java:3603)
        at org.mozilla.javascript.gen._stdin__2.call(Unknown Source)
        at org.mozilla.javascript.gen._stdin__2.exec(Unknown Source)
        at org.mozilla.javascript.tools.shell.Main.processSource(Main.java:497)
        at org.mozilla.javascript.tools.shell.Main.processFiles(Main.java:181)
        at org.mozilla.javascript.tools.shell.Main$IProxy.run(Main.java:101)
        at org.mozilla.javascript.Context.call(Context.java:534)
        at org.mozilla.javascript.ContextFactory.call(ContextFactory.java:525)
        at org.mozilla.javascript.tools.shell.Main.exec(Main.java:163)
        at org.mozilla.javascript.tools.shell.Main.main(Main.java:138)
js: exception from uncaught JavaScript throw: java.lang.OutOfMemoryError: Java heap space
js> 
```

Reply:
I'm find with leaving this untested, but it raises a bigger problem that we have in a lot of places -- JavaScript strings are supposed to have a maximum length of 2^53 but Java strings have a maximum length of 2^31. I don't think that we can change that all over Rhino. But if someone DOES try to create a raw string that's longer than that, then we'll likely leak a runtime exception from somewhere in Java.

I think we're better off here throwing an exception if the length is larger than the maximum integer size rather than let some other kind of exception bubble up.

Reply:
So, if literalSegments > Interger.MAX_VALUE throw a RangeError? And then I could make nextIndex an int and remove this check?

Reply:
Yes! I think that would prevent worse problems later.

Reply:
One other question that I wasn't able to find an answer to by searching...
Is there a negative performance impact for checking `if (a == b)` many times when one variable is an int and another is a long? My understanding is that the int is implicitly converted to a long. Would it be better to do this to avoid the conversions later, or does it not really matter?
```java
long rawLength = NativeArray.getLengthProperty(cx, raw);
if (rawLength > Integer.MAX_VALUE) {/*throw RangeError*/}
int literalSegments = (int) rawLength;
```

Reply:
I went ahead and implemented it the way I asked about in my last question. In addition to removing the implicit conversion in the comparison, it also removed the need to explicitly cast long to int in two places further down in the code.